File: src/main/webapp/app/entities/version/version.service.ts
/*
 * Copyright © 2017-2023 CESSDA ERIC (support@cessda.eu)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable } from '@angular/core';
import { HttpClient, HttpResponse } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import moment from 'moment';

import { DATE_FORMAT } from 'app/shared/constants/input.constants';
import { SERVER_API_URL } from 'app/app.constants';
import { createRequestOption, SearchWithPagination } from 'app/shared/util/request-util';
import { Version } from 'app/shared/model/version.model';

type EntityResponseType = HttpResponse<Version>;
type EntityArrayResponseType = HttpResponse<Version[]>;

@Injectable({ providedIn: 'root' })
export class VersionService {
  public resourceUrl = SERVER_API_URL + 'api/versions';
  public resourceSearchUrl = SERVER_API_URL + 'api/_search/versions';

  constructor(protected http: HttpClient) {}

  create(version: Version): Observable<EntityResponseType> {
    const copy = this.convertDateFromClient(version);
    return this.http
      .post<Version>(this.resourceUrl, copy, { observe: 'response' })
      .pipe(map((res: EntityResponseType) => this.convertDateFromServer(res)));
  }

  update(version: Version): Observable<EntityResponseType> {
    const copy = this.convertDateFromClient(version);
    return this.http
      .put<Version>(this.resourceUrl, copy, { observe: 'response' })
      .pipe(map((res: EntityResponseType) => this.convertDateFromServer(res)));
  }

  find(id: number): Observable<EntityResponseType> {
    return this.http
      .get<Version>(`${this.resourceUrl}/${id}`, { observe: 'response' })
      .pipe(map((res: EntityResponseType) => this.convertDateFromServer(res)));
  }

  query(req?: any): Observable<EntityArrayResponseType> {
    const options = createRequestOption(req);
    return this.http
      .get<Version[]>(this.resourceUrl, { params: options, observe: 'response' })
      .pipe(map((res: EntityArrayResponseType) => this.convertDateArrayFromServer(res)));
  }

  delete(id: number): Observable<HttpResponse<object>> {
    return this.http.delete(`${this.resourceUrl}/${id}`, { observe: 'response' });
  }

  search(req: SearchWithPagination): Observable<EntityArrayResponseType> {
    const options = createRequestOption(req);
    return this.http
      .get<Version[]>(this.resourceSearchUrl, { params: options, observe: 'response' })
      .pipe(map((res: EntityArrayResponseType) => this.convertDateArrayFromServer(res)));
  }

  protected convertDateFromClient(version: Version): Version {
    const copy: Version = Object.assign({}, version, {
      publicationDate:
        version.publicationDate && version.publicationDate.isValid() ? version.publicationDate.format(DATE_FORMAT) : undefined,
      lastModified: version.lastModified && version.lastModified.isValid() ? version.lastModified.toJSON() : undefined,
    });
    return copy;
  }

  protected convertDateFromServer(res: EntityResponseType): EntityResponseType {
    if (res.body) {
      res.body.publicationDate = res.body.publicationDate ? moment(res.body.publicationDate) : undefined;
      res.body.lastModified = res.body.lastModified ? moment(res.body.lastModified) : undefined;
    }
    return res;
  }

  protected convertDateArrayFromServer(res: EntityArrayResponseType): EntityArrayResponseType {
    if (res.body) {
      res.body.forEach((version: Version) => {
        version.publicationDate = version.publicationDate ? moment(version.publicationDate) : undefined;
        version.lastModified = version.lastModified ? moment(version.lastModified) : undefined;
      });
    }
    return res;
  }
}

File: src/main/java/eu/cessda/cvs/service/VersionService.java
/*
 * Copyright © 2017-2023 CESSDA ERIC (support@cessda.eu)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package eu.cessda.cvs.service;

import eu.cessda.cvs.service.dto.VersionDTO;
import eu.cessda.cvs.utils.VersionNumber;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Optional;
import java.util.stream.Stream;

/**
 * Service Interface for managing {@link eu.cessda.cvs.domain.Version}.
 */
public interface VersionService {

    /**
     * Save a version.
     *
     * @param versionDTO the entity to save.
     * @return the persisted entity.
     */
    VersionDTO save(VersionDTO versionDTO);

    /**
     * Get all the versions.
     *
     * @param pageable the pagination information.
     * @return the list of entities.
     */
    Page<VersionDTO> findAll(Pageable pageable);

    /**
     * Get the "id" version.
     *
     * @param id the id of the entity.
     * @return the entity.
     */
    Optional<VersionDTO> findOne(Long id);

    /**
     * Delete the "id" version.
     *
     * @param id the id of the entity.
     */
    void delete(Long id);

    /**
     * Find all versions by vocabularyId
     *
     * @param vocabularyId
     * @return
     */
    Stream<VersionDTO> findAllByVocabulary(Long vocabularyId);

    /**
     * Find all published versions by vocabularyId
     *
     * @param vocabularyId
     * @return
     */
    Stream<VersionDTO> findAllPublishedByVocabulary(Long vocabularyId);

    /**
     * Find all older published version than given versionId, from specific vocabulary and language
     *
     * @param vocabularyId
     * @param languageIso
     * @param versionId
     * @return
     */
    Stream<VersionDTO> findOlderPublishedByVocabularyLanguageId(Long vocabularyId, String languageIso, Long versionId);

    /**
     * Find all versions than given versionId, from specific vocabulary and language
     *
     * @param vocabularyId
     * @param versionNumberSl
     * @return
     */
    Stream<VersionDTO> findAllByVocabularyAndVersionSl(Long vocabularyId, VersionNumber versionNumberSl);

    /**
     * Find all published versions than given versionId, from specific vocabulary and language
     *
     * @param vocabularyId
     * @param versionNumberSl
     * @return
     */
    Stream<VersionDTO> findAllPublishedByVocabularyAndVersionSl(Long vocabularyId, VersionNumber versionNumberSl);

    /**
     * Search for the version corresponding to the query.
     *
     * @param query the query of the search.
     *
     * @param pageable the pagination information.
     * @return the list of entities.
     */
    Page<VersionDTO> search(String query, Pageable pageable);

    /**
     * Get list of versions by URN
     * @param urn
     * @return
     */
    Stream<VersionDTO> findByUrn(String urn);

    /**
     * Get list of versions by URN that starting with
     * @param urn
     * @return
     */
    Stream<VersionDTO> findByUrnStartingWith(String urn);

    /**
     * Get list of used languages by version status
     *
     * @param status
     * @return
     */
    Stream<String> findAllLanguagesByStatus(List<String> status);
}
